io/directory.go

```diff
@@ -394,3 +552,63 @@
 
 	return nil
 }
+
+func (d *UpgradeableDirectory) getDagService() ipld.DAGService {
+	switch v := d.Directory.(type) {
+	case *BasicDirectory:
+		return v.dserv
+	case *HAMTDirectory:
+		return v.dserv
+	default:
+		panic("unknown directory type")
+	}
+}
+
+// RemoveChild implements the `Directory` interface. Used in the case where we wrap
+// a HAMTDirectory that might need to be downgraded to a BasicDirectory. The
+// upgrade path is in AddChild.
+// FIXME: Consider adding some margin in the comparison against HAMTShardingSize
+//  to avoid an eager enumeration at the first opportunity we go below it
+//  (in which case we would have the hard comparison in GetNode() to make
+//  sure we make good on the value). Finding the right margin can be tricky
+//  and very dependent on the use case so it might not be worth it.
+func (d *UpgradeableDirectory) RemoveChild(ctx context.Context, name string) error {
```

Comment:
Moving to https://github.com/ipfs/go-unixfs/issues/105.

Reply:
Since we depend on the implied optimization of having the shards available on disk in repeated queries having repeated checks on the size shouldn't be too costly. (Also the `HAMTDirectory.sizeChange` cache should also alleviate eager checks when we actually haven't reduced the directory size.)

Removing this then.